src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp

```diff
@@ -244,36 +266,90 @@ QueryPlanPtr MergeTreeDataSelectExecutor::readFromParts(
-                continue;
-
-            if (partition_pruner)
+            for (const auto & part : prev_parts)
             {
-                if (partition_pruner->canBePruned(part))
+                if (part_values.find(part->name) == part_values.end())
+                    continue;
+
+                if (part->isEmpty())
+                    continue;
+
+                if (minmax_idx_condition
+                    && !minmax_idx_condition->checkInHyperrectangle(part->minmax_idx.hyperrectangle, data.minmax_idx_column_types)
+                            .can_be_true)
                     continue;
+
+                if (partition_pruner)
+                {
+                    if (partition_pruner->canBePruned(part))
+                        continue;
+                }
+
+                if (max_block_numbers_to_read)
+                {
+                    auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
+                    if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
+                        continue;
+                }
+
+                /// populate UUIDs and exclude ignored parts if enabled
+                if (query_context.getSettingsRef().allow_experimental_query_deduplication && part->uuid != UUIDHelpers::Nil)
+                {
+                    /// Skip the part if its uuid is meant to be excluded
+                    if unlikely (ignored_part_uuids->find(part->uuid))
+                        continue;
+
+                    auto result = temp_part_uuids.insert(part->uuid);
+                    if (!result.second)
+                        throw Exception("Found a part with the same UUID on the same replica.", ErrorCodes::LOGICAL_ERROR);
+                }
+
+                parts.push_back(part);
             }
 
-            if (max_block_numbers_to_read)
+            if unlikely (!temp_part_uuids.empty())
             {
-                auto blocks_iterator = max_block_numbers_to_read->find(part->info.partition_id);
-                if (blocks_iterator == max_block_numbers_to_read->end() || part->info.max_block > blocks_iterator->second)
-                    continue;
+                auto duplicates = query_context.getPartUuids()->add(std::vector<UUID>{temp_part_uuids.begin(), temp_part_uuids.end()});
+                if (!duplicates.empty())
+                {
+                    /// if any duplicates appeared during the first pass, adding them to the exclusion, so they will be skipped on second pass
+                    query_context.getIgnoredPartUuids()->add(duplicates);
+                    return false;
+                }
             }
 
-            parts.push_back(part);
+            return true;
+        };
+
+        /// Process parts that have to be read for a query.
+        auto retry = false;
+        while (!process_parts())
```

Comment:
As far as I understood we can have 2 calls of `process_parts` at most. This loop is extremely confusing.